src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Pipeline/ParameterCapturingPipeline.cs

```diff
@@ -50,33 +50,48 @@ public async Task RunAsync(CancellationToken stoppingToken)
             while (!stoppingToken.IsCancellationRequested)
             {
                 CapturingRequest request = await _requestQueue.Reader.ReadAsync(stoppingToken);
-                if (!TryStartCapturing(request.Payload))
+
+                void onFault(object? sender, InstrumentedMethod faultingMethod)
                 {
-                    continue;
+                    _callbacks.ProbeFault(request.Payload.RequestId, faultingMethod);
                 }
 
-                using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
-                cts.CancelAfter(request.Payload.Duration);
-
                 try
                 {
-                    await request.StopRequest.Task.WaitAsync(cts.Token).ConfigureAwait(false);
-                }
-                catch (OperationCanceledException)
-                {
+                    _probeManager.OnProbeFault += onFault;
 
-                }
+                    if (!await TryStartCapturingAsync(request.Payload, stoppingToken).ConfigureAwait(false))
+                    {
+                        continue;
+                    }
 
-                _probeManager.StopCapturing();
+                    using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+                    cts.CancelAfter(request.Payload.Duration);
 
-                _callbacks.CapturingStop(request.Payload.RequestId);
-                _ = _allRequests.TryRemove(request.Payload.RequestId, out _);
+                    try
+                    {
+                        await request.StopRequest.Task.WaitAsync(cts.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+
+                    }
+
+                    await _probeManager.StopCapturingAsync(stoppingToken).ConfigureAwait(false);
```

Comment:
From an implementation-is-opaque perspective, this method will immediately cancel without doing any work if `stoppingToken` is already cancelled. Is that the behavior you want here? You might want an arbitrary but low timeout in the case that shutdown is requested so that probes can be uninstalled; there may be scenarios where an application is performing reload of the entire service collection and will restart the host without ending the process.